### PR TITLE
Only recompile soy templates that have changed, which should make soy recompilation much more responsive.

### PR DESCRIPTION
**Pull request by nicks**
Hello @kylehg, @ikai, 

Please review the following commits I made in branch 'nicks/lazy'.

5234010c48729ba570bd0aab92f6fc5329526193 (2016-12-01 14:06:09 -0500)
Only recompile soy templates that have changed, which should make soy recompilation much more responsive.

Also, adjust directory structure from path/to/file_locale.soy.js to locale/path/to/file.soy.js,
which is easier to build deps for.

R=@kylehg
R=@ikai

### CODE REVIEW

File: lib/SoyCompiler.js [hunk 1]
```diff
@@ -299,6 +303,25 @@ SoyCompiler.prototype._compileTemplateFilesAndEmit = function (inputDir, files,
   })
 
   cp.on('exit', onExit)
+
+  return deferred.promise.then(function () {
+    var vmTypes = [DEFAULT_VM_CONTEXT]
+    if (options.locales && options.locales.length > 0) {
+      vmTypes = options.locales.concat()
```
Comment (kylehg) [inline lib/SoyCompiler.js]: ?

Comment (kylehg) [inline lib/SoyCompiler.js]: If this is just to clone the array, maybe use `.slice()` or `Array.from()` to be more clear that this isn't a bug?

File: test/test.js [hunk 2]
```diff
@@ -32,14 +33,17 @@ exports.setUp = function (done) {
...
   child_process.spawn = function (prog, args, opts) {
     spawnOpts.push(opts)
+    spawnArgs.push(args)
     return spawn.apply(child_process, arguments)
   }
   done()
 }
 
 exports.tearDown = function (done) {
+  console.error('done')
```
Comment (kylehg) [inline test/test.js]: Should this be a `log`?

**Comment by archfear** [comment]
LGTM

File: lib/SoyCompiler.js (hunk 1 above)
Comment (nicks) [inline lib/SoyCompiler.js]: ok, i added a comment

**Pull request merged**